backends/cuda/ceed-cuda-vec.c

```diff
@@ -0,0 +1,352 @@
+// *****************************************************************************
+static int CeedVectorSetValue_Cuda(CeedVector vec, CeedScalar val) {
+  int ierr;
+  Ceed ceed;
+  ierr = CeedVectorGetCeed(vec, &ceed); CeedChk(ierr);
+  CeedVector_Cuda *data;
+  ierr = CeedVectorGetData(vec, (void*)&data); CeedChk(ierr);
+  CeedInt length;
+  ierr = CeedVectorGetLength(vec, &length); CeedChk(ierr);
+  switch(data->memState){
+  case HOST_SYNC:
+    ierr = HostSetValue(data->h_array, length, val);
+    CeedChk(ierr);
+    break;
+  case NONE_SYNC:
+    /*
+      Handles the case where SetValue is used without SetArray.
+      Default allocation then happens on the GPU.
+    */
+    if (data->d_array==NULL)
+    {
+      ierr = cudaMalloc((void**)&data->d_array_allocated, bytes(vec));
+      CeedChk_Cu(ceed, ierr);
+      data->d_array = data->d_array_allocated;
+    }
+    data->memState = DEVICE_SYNC;
+  case DEVICE_SYNC:
+    ierr = DeviceSetValue(data->d_array, length, val);
+    CeedChk(ierr);
+    break;
+  case BOTH_SYNC:
+    ierr = HostSetValue(data->h_array, length, val);
+    CeedChk(ierr);
+    ierr = DeviceSetValue(data->d_array, length, val);
+    CeedChk(ierr);
+    break;
+  }
+  return 0;
+}
+
+
+// *****************************************************************************
+// * Get read-only access to a vector via the specified mtype memory type
+// * on which to access the array. If the backend uses a different memory type,
+// * this will perform a copy (possibly cached).
+// *****************************************************************************
+static int CeedVectorGetArrayRead_Cuda(const CeedVector vec,
+                                       const CeedMemType mtype,
+                                       const CeedScalar **array) {
+  int ierr;
+  Ceed ceed;
+  ierr = CeedVectorGetCeed(vec, &ceed); CeedChk(ierr);
+  CeedVector_Cuda *data;
+  ierr = CeedVectorGetData(vec, (void*)&data); CeedChk(ierr);
+
+  switch (mtype) {
+  case CEED_MEM_HOST:
+    if(data->h_array==NULL){
+      CeedInt length;
+      ierr = CeedVectorGetLength(vec, &length); CeedChk(ierr);
+      ierr = CeedMalloc(length, &data->h_array_allocated);
+      CeedChk(ierr);
+      data->h_array = data->h_array_allocated;
+    }
+    if(data->memState==DEVICE_SYNC){
+      ierr = CeedSyncD2H_Cuda(vec); CeedChk(ierr);
+      data->memState = BOTH_SYNC;
+    }
+    *array = data->h_array;
+    break;
+  case CEED_MEM_DEVICE:
+    if (data->d_array==NULL)
+    {
+      ierr = cudaMalloc((void**)&data->d_array_allocated, bytes(vec));
+      CeedChk_Cu(ceed, ierr);
+      data->d_array = data->d_array_allocated;
+    }
+    if (data->memState==HOST_SYNC){
+      ierr = CeedSyncH2D_Cuda(vec); CeedChk(ierr);
+      data->memState = BOTH_SYNC;
+    }
+    *array = data->d_array;
+    break;
+  }
+  return 0;
+}
+
+// *****************************************************************************
+static int CeedVectorGetArray_Cuda(const CeedVector vec,
```

Comment:
Is this code different from GetArrayRead? I can't see a difference, but I could be wrong. If it's the same we can just use the same function for both.

Reply:
It's not the same, the synchronisation logic is not the same.

Reply:
`GetArrayRead` sync results in `BOTH_SYNC` whereas `GetArray` never results in `BOTH_SYNC` since the array can be modified.